TST: test_replace_mmap_data tests fail on Windows

On Appveyor CI, `test_replace_mmap_data` and `test_replace_mmap_data_2` failed with the following errors. `git blame` indicated that the affected tests and code have not changed for 5-6 years (I might be wrong, so please double check), so chances are something changed on the Appveyor side? How should we proceed?

```
================================== FAILURES ===================================
_________________ TestHDUListFunctions.test_replace_mmap_data _________________
self = <astropy.io.fits.tests.test_hdulist.TestHDUListFunctions object at 0x0000028347ECC780>
    def test_replace_mmap_data(self):
        """Regression test for
        https://github.com/spacetelescope/PyFITS/issues/25
    
        Replacing the mmap'd data of one file with mmap'd data from a
        different file should work.
        """
    
        arr_a = np.arange(10)
        arr_b = arr_a * 2
    
        def test(mmap_a, mmap_b):
            hdu_a = fits.PrimaryHDU(data=arr_a)
            hdu_a.writeto(self.temp('test_a.fits'), overwrite=True)
            hdu_b = fits.PrimaryHDU(data=arr_b)
            hdu_b.writeto(self.temp('test_b.fits'), overwrite=True)
    
            hdul_a = fits.open(self.temp('test_a.fits'), mode='update',
                               memmap=mmap_a)
            hdul_b = fits.open(self.temp('test_b.fits'), memmap=mmap_b)
            hdul_a[0].data = hdul_b[0].data
            hdul_a.close()
            hdul_b.close()
    
            hdul_a = fits.open(self.temp('test_a.fits'))
    
            assert np.all(hdul_a[0].data == arr_b)
    
        with ignore_warnings():
            test(True, True)
    
            # Repeat the same test but this time don't mmap A
>           test(False, True)
astropy\io\fits\tests\test_hdulist.py:814: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
astropy\io\fits\tests\test_hdulist.py:795: in test
    hdu_a.writeto(self.temp('test_a.fits'), overwrite=True)
astropy\utils\decorators.py:503: in wrapper
    return function(*args, **kwargs)
astropy\io\fits\hdu\base.py:373: in writeto
    checksum=checksum)
astropy\utils\decorators.py:503: in wrapper
    return function(*args, **kwargs)
astropy\io\fits\hdu\hdulist.py:919: in writeto
    fileobj = _File(fileobj, mode=mode, overwrite=overwrite)
astropy\utils\decorators.py:503: in wrapper
    return function(*args, **kwargs)
astropy\io\fits\file.py:178: in __init__
    self._open_filename(fileobj, mode, overwrite)
astropy\io\fits\file.py:544: in _open_filename
    self._overwrite_existing(overwrite, None, True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
self = <[AttributeError("'_File' object has no attribute '_file'") raised in repr()] _File object at 0x28346ed0390>
overwrite = True, fileobj = None, closed = True
    def _overwrite_existing(self, overwrite, fileobj, closed):
        """Overwrite an existing file if ``overwrite`` is ``True``, otherwise
        raise an OSError.  The exact behavior of this method depends on the
        _File object state and is only meant for use within the ``_open_*``
        internal methods.
        """
    
        # The file will be overwritten...
        if ((self.file_like and hasattr(fileobj, 'len') and fileobj.len > 0) or
            (os.path.exists(self.name) and os.path.getsize(self.name) != 0)):
            if overwrite:
                if self.file_like and hasattr(fileobj, 'truncate'):
                    fileobj.truncate(0)
                else:
                    if not closed:
                        fileobj.close()
>                   os.remove(self.name)
E                   PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\fits-test-5dy_4ocw\\test_a.fits'
astropy\io\fits\file.py:436: PermissionError
________________ TestHDUListFunctions.test_replace_mmap_data_2 ________________
self = <astropy.io.fits.tests.test_hdulist.TestHDUListFunctions object at 0x0000028347B50B00>
    def test_replace_mmap_data_2(self):
        """Regression test for
        https://github.com/spacetelescope/PyFITS/issues/25
    
        Replacing the mmap'd data of one file with mmap'd data from a
        different file should work.  Like test_replace_mmap_data but with
        table data instead of image data.
        """
    
        arr_a = np.arange(10)
        arr_b = arr_a * 2
    
        def test(mmap_a, mmap_b):
            col_a = fits.Column(name='a', format='J', array=arr_a)
            col_b = fits.Column(name='b', format='J', array=arr_b)
            hdu_a = fits.BinTableHDU.from_columns([col_a])
            hdu_a.writeto(self.temp('test_a.fits'), overwrite=True)
            hdu_b = fits.BinTableHDU.from_columns([col_b])
            hdu_b.writeto(self.temp('test_b.fits'), overwrite=True)
    
            hdul_a = fits.open(self.temp('test_a.fits'), mode='update',
                               memmap=mmap_a)
            hdul_b = fits.open(self.temp('test_b.fits'), memmap=mmap_b)
            hdul_a[1].data = hdul_b[1].data
            hdul_a.close()
            hdul_b.close()
    
            hdul_a = fits.open(self.temp('test_a.fits'))
    
            assert 'b' in hdul_a[1].columns.names
            assert 'a' not in hdul_a[1].columns.names
            assert np.all(hdul_a[1].data['b'] == arr_b)
    
        with ignore_warnings():
            test(True, True)
    
            # Repeat the same test but this time don't mmap A
            test(False, True)
    
            # Finally, without mmaping B
>           test(True, False)
astropy\io\fits\tests\test_hdulist.py:859: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
astropy\io\fits\tests\test_hdulist.py:835: in test
    hdu_a.writeto(self.temp('test_a.fits'), overwrite=True)
astropy\utils\decorators.py:503: in wrapper
    return function(*args, **kwargs)
astropy\io\fits\hdu\base.py:1543: in writeto
    checksum=checksum)
astropy\utils\decorators.py:503: in wrapper
    return function(*args, **kwargs)
astropy\io\fits\hdu\hdulist.py:919: in writeto
    fileobj = _File(fileobj, mode=mode, overwrite=overwrite)
astropy\utils\decorators.py:503: in wrapper
    return function(*args, **kwargs)
astropy\io\fits\file.py:178: in __init__
    self._open_filename(fileobj, mode, overwrite)
astropy\io\fits\file.py:544: in _open_filename
    self._overwrite_existing(overwrite, None, True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
self = <[AttributeError("'_File' object has no attribute '_file'") raised in repr()] _File object at 0x28347d6bdd8>
overwrite = True, fileobj = None, closed = True
    def _overwrite_existing(self, overwrite, fileobj, closed):
        """Overwrite an existing file if ``overwrite`` is ``True``, otherwise
        raise an OSError.  The exact behavior of this method depends on the
        _File object state and is only meant for use within the ``_open_*``
        internal methods.
        """
    
        # The file will be overwritten...
        if ((self.file_like and hasattr(fileobj, 'len') and fileobj.len > 0) or
            (os.path.exists(self.name) and os.path.getsize(self.name) != 0)):
            if overwrite:
                if self.file_like and hasattr(fileobj, 'truncate'):
                    fileobj.truncate(0)
                else:
                    if not closed:
                        fileobj.close()
>                   os.remove(self.name)
E                   PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\fits-test-88vh7_bv\\test_a.fits'
astropy\io\fits\file.py:436: PermissionError
===== 2 failed, 11833 passed, 338 skipped, 110 xfailed in 236.77 seconds ======
```

This was first reported by @astrofrog in Slack, who also pinged @saimn .

--- a/astropy/io/fits/tests/test_hdulist.py
+++ b/astropy/io/fits/tests/test_hdulist.py
@@ -796,16 +796,16 @@ class TestHDUListFunctions(FitsTestCase):
             hdu_b = fits.PrimaryHDU(data=arr_b)
             hdu_b.writeto(self.temp('test_b.fits'), overwrite=True)
 
-            hdul_a = fits.open(self.temp('test_a.fits'), mode='update',
-                               memmap=mmap_a)
-            hdul_b = fits.open(self.temp('test_b.fits'), memmap=mmap_b)
-            hdul_a[0].data = hdul_b[0].data
-            hdul_a.close()
-            hdul_b.close()
+            with fits.open(self.temp('test_a.fits'), mode='update',
+                           memmap=mmap_a) as hdul_a:
+                with fits.open(self.temp('test_b.fits'),
+                               memmap=mmap_b) as hdul_b:
+                    hdul_a[0].data = hdul_b[0].data
 
-            hdul_a = fits.open(self.temp('test_a.fits'))
+            with fits.open(self.temp('test_a.fits')) as hdul_a:
+                status = np.all(hdul_a[0].data == arr_b)
 
-            assert np.all(hdul_a[0].data == arr_b)
+            assert status
 
         with ignore_warnings():
             test(True, True)
@@ -836,18 +836,18 @@ class TestHDUListFunctions(FitsTestCase):
             hdu_b = fits.BinTableHDU.from_columns([col_b])
             hdu_b.writeto(self.temp('test_b.fits'), overwrite=True)
 
-            hdul_a = fits.open(self.temp('test_a.fits'), mode='update',
-                               memmap=mmap_a)
-            hdul_b = fits.open(self.temp('test_b.fits'), memmap=mmap_b)
-            hdul_a[1].data = hdul_b[1].data
-            hdul_a.close()
-            hdul_b.close()
+            with fits.open(self.temp('test_a.fits'), mode='update',
+                           memmap=mmap_a) as hdul_a:
+                with fits.open(self.temp('test_b.fits'),
+                               memmap=mmap_b) as hdul_b:
+                    hdul_a[1].data = hdul_b[1].data
 
-            hdul_a = fits.open(self.temp('test_a.fits'))
+            with fits.open(self.temp('test_a.fits')) as hdul_a:
+                status = (('b' in hdul_a[1].columns.names) and
+                          ('a' not in hdul_a[1].columns.names) and
+                          np.all(hdul_a[1].data['b'] == arr_b))
 
-            assert 'b' in hdul_a[1].columns.names
-            assert 'a' not in hdul_a[1].columns.names
-            assert np.all(hdul_a[1].data['b'] == arr_b)
+            assert status
 
         with ignore_warnings():
             test(True, True)

--- a/astropy/io/fits/tests/test_hdulist.py
+++ b/astropy/io/fits/tests/test_hdulist.py
@@ -803,9 +803,7 @@ class TestHDUListFunctions(FitsTestCase):
                     hdul_a[0].data = hdul_b[0].data
 
             with fits.open(self.temp('test_a.fits')) as hdul_a:
-                status = np.all(hdul_a[0].data == arr_b)
-
-            assert status
+                assert np.all(hdul_a[0].data == arr_b)
 
         with ignore_warnings():
             test(True, True)
@@ -843,11 +841,9 @@ class TestHDUListFunctions(FitsTestCase):
                     hdul_a[1].data = hdul_b[1].data
 
             with fits.open(self.temp('test_a.fits')) as hdul_a:
-                status = (('b' in hdul_a[1].columns.names) and
-                          ('a' not in hdul_a[1].columns.names) and
-                          np.all(hdul_a[1].data['b'] == arr_b))
-
-            assert status
+                assert 'b' in hdul_a[1].columns.names
+                assert 'a' not in hdul_a[1].columns.names
+                assert np.all(hdul_a[1].data['b'] == arr_b)
 
         with ignore_warnings():
             test(True, True)

